control-plane/aegisops_control_plane/adapters/postgres.py

```
@@ -526,6 +526,27 @@ def list(self, record_type: Type[RecordT]) -> tuple[RecordT, ...]:
             for row in rows
         )
 
+    def latest_ai_trace_record(self) -> AITraceRecord | None:
+        table = self._table_config(AITraceRecord)
+        query = (
+            f"select {', '.join(table.record_fields)} "
+            f"from aegisops_control.{table.table_name} "
+            "order by generated_at desc, ai_trace_id desc limit 1"
```

Ruff (0.15.11)

[error] Possible SQL injection vector through string-based query construction

(S608)